NorthstarDLL/server/servernethooks.cpp

```diff
@@ -0,0 +1,218 @@
+		return;
+	}
+
+	std::string pSig; // is binary data, not actually an ASCII string
+	pSig = pData.substr(0, HMACSHA256_LEN);
+	pData = pData.substr(HMACSHA256_LEN);
+
+	if (!g_pMasterServerManager || !g_pMasterServerManager->m_sOwnServerAuthToken[0])
+	{
+		if (dbg)
+			spdlog::warn(
+				"ignoring Atlas connectionless packet (size={} type={}): invalid (data={}): no masterserver token yet",
+				packet->size,
+				pType,
+				pData);
+		return;
+	}
+
+	if (!VerifyHMACSHA256(std::string(g_pMasterServerManager->m_sOwnServerAuthToken), pSig, pData))
+	{
+		if (!Cvar_net_debug_atlas_packet_insecure->GetBool())
+		{
+			if (dbg)
+				spdlog::warn(
+					"ignoring Atlas connectionless packet (size={} type={}): invalid: invalid signature (key={})",
+					packet->size,
+					pType,
+					std::string(g_pMasterServerManager->m_sOwnServerAuthToken));
+			return;
+		}
+		spdlog::warn(
+			"processing Atlas connectionless packet (size={} type={}) with invalid signature due to net_debug_atlas_packet_insecure",
+			packet->size,
+			pType);
+	}
+
+	if (dbg)
+		spdlog::info("got Atlas connectionless packet (size={} type={} data={})", packet->size, pType, pData);
+
+	std::thread t(&MasterServerManager::ProcessConnectionlessPacketSigreq1, g_pMasterServerManager, pData);
+	t.detach();
+
+	return;
+}
+
+static void ProcessAtlasConnectionlessPacket(R2::netpacket_t* packet)
+{
+	bool dbg = Cvar_net_debug_atlas_packet->GetBool();
+
+	// extract kind, null-terminated type, data
+	std::string pType, pData;
+	for (int i = 5; i < packet->size; i++)
+	{
+		if (packet->data[i] == '\x00')
+		{
+			pType.assign((char*)(&packet->data[5]), (size_t)(i - 5));
+			if (i + 1 < packet->size)
+				pData.assign((char*)(&packet->data[i + 1]), (size_t)(packet->size - i - 1));
+			break;
+		}
+	}
+
+	// note: all Atlas connectionless packets should be idempotent so multiple attempts can be made to mitigate packet loss
+	// note: all long-running Atlas connectionless packet handlers should be started in a new thread (with copies of the data) to avoid
+	// blocking networking
+
+	// v1 HMACSHA256-signed masterserver request
+	if (pType == "sigreq1")
+	{
+		ProcessAtlasConnectionlessPacketSigreq1(packet, dbg, pType, pData);
+		return;
+	}
+
+	if (dbg)
+		spdlog::warn("ignoring Atlas connectionless packet (size={} type={}): unknown type", packet->size, pType);
+	return;
+}
+
+AUTOHOOK(ProcessConnectionlessPacket, engine.dll + 0x117800, bool, , (void* a1, R2::netpacket_t* packet))
+{
+	// packet->data consists of 0xFFFFFFFF (int32 -1) to indicate packets aren't split, followed by a header consisting of a single
+	// character, which is used to uniquely identify the packet kind. Most kinds follow this with a null-terminated string payload
+	// then an arbitrary amoount of data.
+
+	// T (no rate limits since we authenticate packets before doing anything expensive)
+	if (4 < packet->size && packet->data[4] == 'T')
```

Comment:
a constant rather than just 'T' here would be nice, but not required

Reply:
figured it was more clear to be able to see all the types in one function